tests/testthat/test-missing_data.R

```diff
@@ -5,31 +5,139 @@ test_that("defMiss produces correct output", {
   baseline <- TRUE
   monotonic <- FALSE
-
+  
   def <- data.table::data.table(
     varname, formula, logit.link,
     baseline, monotonic
   )
-
+  
   expect_equal(def, defMiss(
     NULL, varname, formula, logit.link,
     baseline, monotonic
   ))
-
+  
   expect_equal(rbind(def, def), defMiss(
     def, varname, formula,
     logit.link, baseline, monotonic
   ))
 })
 
+# genMiss ----
+test_that("genMiss works", {
+  # not repeated
+  def1 <- defData(varname = "m", dist = "binary", formula = .5)
+  def1 <- defData(def1, "u", dist = "binary", formula = .5)
+  def1 <- defData(def1, "x1", dist="normal", formula = "20*m + 20*u", variance = 2)
+  def1 <- defData(def1, "x2", dist="normal", formula = "20*m + 20*u", variance = 2)
+  def1 <- defData(def1, "x3", dist="normal", formula = "20*m + 20*u", variance = 2)
+  
+  dtAct1 <- genData(7500, def1)
+  
+  hardProbForm <- runif(1)
+  form1val0 <- runif(1)
+  form1val1 <- runif(1)
+  form1 <- "..form1val0*(1-m) + m*..form1val1"
+  #form1 <- paste0(form1val0, "*(1-m) + m*", form1val1)
+  
+  defM <- defMiss(varname = "x1", formula = hardProbForm, logit.link = FALSE)
+  defM <- defMiss(defM, varname = "x2", formula = form1, logit.link = FALSE)
+  defM <- defMiss(defM, varname = "u", formula = 1, logit.link = FALSE) # not observed
+  
+  missMat <- genMiss(dtAct1, defM, idvars = "id")
+  
+  ## check probabilistic column has correct distribution
+  hardProbAct <- mean(missMat[, x1])
+  hPDiff <- abs(hardProbAct - hardProbForm)
+  
+  expect_true(hPDiff < 0.02)
+  
+  ## check dependent column 1
+  #check probs when m is 0/1, u is 0/1
+  
+  mIs0 <- dtAct1[, m == 0]
+  mIs1 <- dtAct1[, m == 1]
+
+  missMat0 <- missMat[mIs0]
+  missMat1 <- missMat[mIs1]
+  
+  avg0 <- mean(missMat0[, x2])
+  avg1 <- mean(missMat1[, x2])
+  
+  diff0 <- abs(form1val0 -avg0)
+  diff1 <- abs(form1val1 -avg1)
+  
+  expect_true(diff0 < 0.02 && diff1 < 0.02)
+  
+  ## check empty column
+  expect_true(all(missMat[, u] == 1))
+  
+  # repeated (not worth looking at if !baseline and !mono)
+  ## includes lags
+  
+  ## baseline
+  #if missing at baseline, missing at all other perios
+  idNum <- 500
+  
+  dtAct2 <- genData(idNum, def1)
+  dtAct2 <- trtObserve(dtAct2, formulas = .5, logit.link = FALSE, grpName = "rx")
+  
+  defLong <- defDataAdd(varname = "y", dist = "normal", formula = "10 + period*2 + 2 * rx", variance = 2)
+  
+  dtTime <- addPeriods(dtAct2, nPeriods = 4)
+  dtTime <- addColumns(defLong, dtTime)
+  
+  defMlong <- defMiss(varname = "x1", formula = .20, baseline = TRUE)
+  defMlong <- defMiss(defMlong,varname = "y", formula = "-1.5 - 1.5 * rx + .25*period", logit.link = TRUE, baseline = FALSE, monotonic = FALSE)
+  
+  missMatLong <- genMiss(dtTime, defMlong, idvars = c("id","rx"), repeated = TRUE, periodvar = "period")
+  
+  casevec <- NULL
+  for(i in 1:idNum) {
+    tempMM <- missMatLong[id == i]
+    if(tempMM[1, x1] == 1) {
+      casevec <- c(casevec, all(tempMM[, x1] == 1))
+    }
+  }
```

Comment:
In general, `for` loops in R are super slow, so do just about anything to avoid them (unless you know they will loop a small number of times). This is an alternative, speedier option (anything with data.tables is going to be fast, because data.table is super fast):

```
ids <- missMatLong[period == 0 & x1 == 1, id]
casevec <- missMatLong[id %in% ids, all(x1 == 1)]
```

Also, do we check for the case where `x1` might be 0 at baseline (time 0), implying that all others must be 0 as well?

Reply:
I'll implement it this way.

Ah I didn't realize that case was also true with baseline, I'll put that in.